File: 7/712/MinimumASCIIDeleteSumForTwoStrings.java
/*
https://leetcode.com/problems/minimum-ascii-delete-sum-for-two-strings/description/?envType=daily-question&envId=2026-01-10
*/

class Solution {
    public int minimumDeleteSum(String s1, String s2) {
        int lengthOfS1=s1.length(), lengthOfS2=s2.length();
        int[][]dp=new int[lengthOfS1+1][lengthOfS2+1];
        for(int i=0; i<=lengthOfS1; ++i){
            for(int j=0; j<=lengthOfS2; ++j){
                dp[i][j]=0;
            }
        }
        for(int i=lengthOfS1-1; i>=0; --i){
            dp[i][lengthOfS2]=s1.charAt(i)+dp[i+1][lengthOfS2];
        }
        for(int j=lengthOfS2-1; j>=0; --j){
            dp[lengthOfS1][j]=s2.charAt(j)+dp[lengthOfS1][j+1];
        }
        for(int i=lengthOfS1-1; i>=0; --i){
            for(int j=lengthOfS2-1; j>=0; --j){
                if(s1.charAt(i)==s2.charAt(j)){
                    dp[i][j]=dp[i+1][j+1];
                } else{
                    dp[i][j]=Math.min(s1.charAt(i)+dp[i+1][j], s2.charAt(j)+dp[i][j+1]);
                }
            }
        }
        return dp[0][0];
    }
}

File: 7/712/MinimumASCIIDeleteSumForTwoStrings.cpp
/*
https://leetcode.com/problems/minimum-ascii-delete-sum-for-two-strings/description/?envType=daily-question&envId=2026-01-10
*/

class Solution {
public:
    int minimumDeleteSum(string s1, string s2) {
        int sz1=s1.size(), sz2=s2.size();
        vector<vector<int>>dp(sz1+1, vector<int>(sz2+1, 0));
        for(int i=sz1-1; i>=0; --i){
            dp[i][sz2]=s1[i]+dp[i+1][sz2];
        }
        for(int j=sz2-1; j>=0; --j){
            dp[sz1][j]=s2[j]+dp[sz1][j+1];
        }
        for(int i=sz1-1; i>=0; --i){
            for(int j=sz2-1; j>=0; --j){
                if(s1[i]==s2[j]){
                    dp[i][j]=dp[i+1][j+1];
                } else{
                    dp[i][j]=min(s1[i]+dp[i+1][j], s2[j]+dp[i][j+1]);
                }
            }
        }
        return dp[0][0];
    }
};